torch_geometric/nn/models/g_retriever.py

```diff
@@ -0,0 +1,328 @@
+                num_layers=num_gnn_layers,
+                norm='batch_norm',
+            ).to(self.llm_device)
+        # For the MLP Projection
+        mlp_hidden_dim = gnn_out_channels
+        self.projector = nn.Sequential(
+            nn.Linear(gnn_out_channels, mlp_hidden_dim),
+            nn.Sigmoid(),
+            nn.Linear(mlp_hidden_dim, mlp_out_dim),
+        ).to(self.llm_device)
+
+        self.word_embedding = self.llm_to_use.word_embedding
+
+    def encode_graphs(self, node_feat, edge_index, edge_attr, batch):
+        x = node_feat.to(self.llm_device)
+        edge_index = edge_index.long().to(self.llm_device)
+        edge_attr = edge_attr.to(self.llm_device)
+        n_embeds = self.graph_encoder(x, edge_index.long(), edge_attr)
+        batch = batch.to(self.llm_device)
+        # mean pooling
+        g_embeds = scatter(n_embeds, batch, dim=0, reduce='mean')
+        return g_embeds
+
+    def forward(
+        self,
+        question: List[str],
+        node_feat: torch.Tensor,
+        edge_index: torch.Tensor,
+        batch: torch.Tensor,
+        ptr: torch.Tensor,
+        label: List[str],
+        edge_attr: Optional[torch.Tensor] = None,
+        additional_text_context: Optional[List[str]] = None,
+    ):
+        r"""Forward pass.
+
+        Args:
+            question (List[str]): The questions/prompts.
+            node_feat (torch.Tensor): The input node features.
+            edge_index (torch.Tensor or SparseTensor): The edge indices.
+            batch (torch.Tensor): The batch vector
+                :math:`\mathbf{b} \in {\{ 0, \ldots, B-1\}}^N`, which assigns
+                each element to a specific example.
+            ptr (torch.Tensor): The pointer vector, denoting the
+                boundaries between examples in the batch.
+            label (List[str]): The answers/labels.
+            edge_attr (torch.Tensor, optional): The edge features (if supported
+                by the GNN being used). (default: :obj:`None`)
+            additional_text_context (List[str], optional): Additional context
+                to give to the LLM, such as textified knowledge graphs.
+        """
+        batch_size, questions, context, eos_user_tokens, \
+            bos_embeds, pad_embeds = self.llm_to_use._encode_inputs(question, additional_text_context) # noqa
+        # encode labels
+        labels = self.tokenizer(label, add_special_tokens=False)
+        # encode training specific special token
+        eos_tokens = self.tokenizer(EOS, add_special_tokens=False)
+
+        # encode graphs
+        graph_embeds = self.encode_graphs(node_feat, edge_index, edge_attr,
+                                          batch)
+        graph_embeds = self.projector(graph_embeds)
+        batch_inputs_embeds = []
+        batch_attention_mask = []
+        batch_label_input_ids = []
+        num_nodes_per_graph = ptr[1:] - ptr[:-1]
+        for i in range(batch_size):
```

Comment:
There seems to be a lot of duplication here and in `LLM`. Why was it done this way? Is there a way to de-duplicate the amount of code here?